proposals/p5337.md

```diff
@@ -0,0 +1,1057 @@
+interface, we change the existing usage of `extend` from being an introducer of
+its own declaration, to a modifier on a `require` declaration. In this case, the
+only type that can be constrained is `Self`, since that is the type being
+extended. To match the restriction that an `extend impl` must be followed by
+`as` in `class` scope, `extend require` can only be followed by `impls` in
+`interface` scope:
+
+-   `extend require Self impls I` ❌ forbidden, `Self` omitted since no other
+    value allowed
+-   `extend require impls I` ✅ allowed
+-   `require T impls I` ❌ forbidden
+
+`extend require impls I` after this proposal has the same meaning as `extend I`
+did before this proposal, so:
+
+```carbon
+interface A {
+  fn F();
+}
+
+interface B {
+  extend require impls A;
+  fn G();
+}
+```
+
+is equivalent to:
+
+```carbon
+interface A {
+  fn F();
+}
+
+interface B {
+  require impls A;
+  alias F = A.F;
+  fn G();
+}
+```
+
+We then add another way to use `extend` as a modifier in an `interface` scope.
+We define the meaning of `extend impl as I;` in an `interface J` scope to do two
+things:
+
+-   Copy the members of `I` to form new members of `J`.
+-   Define a blanket `impl` that anything that `impls J` also `impls I`, by
+    forwarding to the corresponding members of `J`.
+
+These semantics are intended to be similar to `extend impl` in a `class`, except
+that the definition of the blanket `impl` is generated automatically (and `impl`
+without `extend` is not allowed in an interface).
+
+In this example:
+
+```carbon
+interface I {
+  let T:! type;
+  fn F();
+  fn G();
+}
+
+interface J {
+  extend impl as I;
+  fn H();
+}
+```
+
+The definition of `interface J` is equivalent to:
+
+```carbon
+interface J {
+  let T:! type;
+  fn F();
+  fn G();
+  fn H();
+}
+impl forall [U:! J] U as I {
+  let T:! type = U.(J.T);
+  fn F() = U.(J.F);
+  fn G() = U.(J.G);
+}
+```
+
+The keyword `final` can be added between `extend` and `impl` to make the
+generated `impl` definition `final`. For example, the
+`final impl forall [T:! type, U:! ImplicitAs(T)] U as As(T)` from
+[the "Problem" section](#problem) would be generated from this:
+
+```carbon
+interface ImplicitAs(T:! type) {
+  extend final impl as As(T);
+}
+```
+
+To allow a non-final `impl` to overlap a final `impl`, we say the final `impl`
+is prioritized over a non-final `impl` anytime they overlap. This addresses a
+[problem](#problem) caused by the fact that we don't support the negative
+constraints that would allow a developer to avoid the overlap -- and the only
+sound choice is to prefer the final `impl`. However, if a non-final `impl` can
+never be used because it is completely subsumed by a final `impl`, that is an
+error.
+
+To allow multiple interfaces extending the same interface, we also want to
+support a way to declare overlapping final `impl`s. In this case, list the
+`impl` names in a `final match_first` block instead of putting `final` on the
```

Comment:
This does not mention that final impls can overlap if one comes from the root self type library and one from the interface library (and they are different libraries).